File: creational/factory_method/document_processor/typescript/document_processor.ts
import * as fs from "fs";
import * as path from "path";

// --- Product Interface ---
interface Document {
  title: string;
  content: string[];
  save(filePathBase: string): Promise<void>; // Changed signature, now async
}

// --- Concrete Products ---
class TextDocument implements Document {
  constructor(public title: string, public content: string[]) {
    console.log(
      `TextDocument created: Title='${this.title}', Content preview='${
        content[0] ?? ""
      }...'`
    );
  }

  async save(filePathBase: string): Promise<void> {
    const fullPath = `${filePathBase}.txt`;
    console.log(`Saving Text document to: ${fullPath}`);
    const fileContent = `Title: ${this.title}\n${"=".repeat(
      this.title.length + 2
    )}\n\n${this.content.join("\n")}`;
    try {
      await fs.promises.writeFile(fullPath, fileContent, "utf-8");
      console.log("Text document saved successfully.");
    } catch (err) {
      console.error(`Error saving text document ${fullPath}:`, err);
      throw err; // Re-throw error for testability/handling
    }
  }
}

class JSONDocument implements Document {
  constructor(public title: string, public content: string[]) {
    console.log(
      `JSONDocument created: Title='${this.title}', Content preview='${
        content[0] ?? ""
      }...'`
    );
  }

  async save(filePathBase: string): Promise<void> {
    const fullPath = `${filePathBase}.json`;
    console.log(`Saving JSON document to: ${fullPath}`);
    const data = {
      title: this.title,
      content: this.content,
    };
    try {
      await fs.promises.writeFile(
        fullPath,
        JSON.stringify(data, null, 4),
        "utf-8"
      );
      console.log("JSON document saved successfully.");
    } catch (err) {
      console.error(`Error saving JSON document ${fullPath}:`, err);
      throw err;
    }
  }
}

class HTMLDocument implements Document {
  constructor(public title: string, public content: string[]) {
    console.log(
      `HTMLDocument created: Title='${this.title}', Content preview='${
        content[0] ?? ""
      }...'`
    );
  }

  async save(filePathBase: string): Promise<void> {
    const fullPath = `${filePathBase}.html`;
    console.log(`Saving HTML document to: ${fullPath}`);
    let htmlContent = `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>${this.title}</title>
    <style>
        body { font-family: sans-serif; line-height: 1.6; padding: 20px; }
        h1 { color: #333; }
        p { margin-bottom: 10px; }
    </style>
</head>
<body>
    <h1>${this.title}</h1>
`;
    this.content.forEach((line) => {
      // Basic paragraph wrapping, consider escaping HTML chars in real use
      htmlContent += `    <p>${line}</p>\n`;
    });
    htmlContent += `</body>\n</html>\n`;

    try {
      await fs.promises.writeFile(fullPath, htmlContent, "utf-8");
      console.log("HTML document saved successfully.");
    } catch (err) {
      console.error(`Error saving HTML document ${fullPath}:`, err);
      throw err;
    }
  }
}

// --- Creator Interface ---
abstract class DocumentProcessor {
  // The Factory Method
  public abstract createDocument(title: string, content: string[]): Document;

  // Core logic using the factory method
  public async processAndSave(
    title: string,
    content: string[],
    outputDir: string,
    filenameBase: string
  ): Promise<void> {
    // Ensure output directory exists
    try {
      await fs.promises.mkdir(outputDir, { recursive: true });
    } catch (err) {
      console.error(`Error creating directory ${outputDir}:`, err);
      return; // Exit if directory creation fails
    }

    // Create the document using the factory method
    const doc = this.createDocument(title, content);

    // Construct the full path without extension
    const fullPathBase = path.join(outputDir, filenameBase);

    // Save the document
    console.log(`\nProcessing with ${this.constructor.name}:`);
    try {
      await doc.save(fullPathBase);
    } catch (err) {
      console.error(
        `Failed to save document ${filenameBase} using ${this.constructor.name}.`
      );
      // Decide if processing should stop or continue
    }
  }
}

// --- Concrete Creators ---
class TextProcessor extends DocumentProcessor {
  public createDocument(title: string, content: string[]): TextDocument {
    console.log("TextProcessor: Creating TextDocument.");
    return new TextDocument(title, content);
  }
}

class JSONProcessor extends DocumentProcessor {
  public createDocument(title: string, content: string[]): JSONDocument {
    console.log("JSONProcessor: Creating JSONDocument.");
    return new JSONDocument(title, content);
  }
}

class HTMLProcessor extends DocumentProcessor {
  public createDocument(title: string, content: string[]): HTMLDocument {
    console.log("HTMLProcessor: Creating HTMLDocument.");
    return new HTMLDocument(title, content);
  }
}

// Export classes and interfaces
export {
  Document,
  TextDocument,
  JSONDocument,
  HTMLDocument,
  DocumentProcessor,
  TextProcessor,
  JSONProcessor,
  HTMLProcessor,
};

File: creational/factory_method/document_processor/python/document_processor.py
from abc import ABC, abstractmethod
import json
import os
from typing import List, Dict, Any

# --- Product Interface ---
class Document(ABC):
    """Abstract Product: Defines the interface for documents."""
    def __init__(self, title: str, content: List[str]):
        self.title = title
        self.content = content
        print(f"Document created: Title='{self.title}', Content preview='{content[0] if content else ''}...'")

    @abstractmethod
    def save(self, path: str):
        """Saves the document to the specified path in its format."""
        pass

# --- Concrete Products ---
class TextDocument(Document):
    """Concrete Product: Implements a plain text document."""
    def save(self, path: str):
        full_path = f"{path}.txt"
        print(f"Saving Text document to: {full_path}")
        try:
            with open(full_path, 'w', encoding='utf-8') as f:
                f.write(f"Title: {self.title}\n")
                f.write("=" * len(self.title) + "==\n\n")
                for line in self.content:
                    f.write(f"{line}\n")
            print("Text document saved successfully.")
        except IOError as e:
            print(f"Error saving text document {full_path}: {e}")

class JSONDocument(Document):
    """Concrete Product: Implements a JSON document."""
    def save(self, path: str):
        full_path = f"{path}.json"
        print(f"Saving JSON document to: {full_path}")
        data: Dict[str, Any] = {
            "title": self.title,
            "content": self.content
        }
        try:
            with open(full_path, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=4)
            print("JSON document saved successfully.")
        except IOError as e:
            print(f"Error saving JSON document {full_path}: {e}")
        except TypeError as e:
            print(f"Error serializing data to JSON for {full_path}: {e}")

class HTMLDocument(Document):
    """Concrete Product: Implements a simple HTML document."""
    def save(self, path: str):
        full_path = f"{path}.html"
        print(f"Saving HTML document to: {full_path}")
        html_content = f"""<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>{self.title}</title>
    <style>
        body {{ font-family: sans-serif; line-height: 1.6; padding: 20px; }}
        h1 {{ color: #333; }}
        p {{ margin-bottom: 10px; }}
    </style>
</head>
<body>
    <h1>{self.title}</h1>
"""
        for line in self.content:
            # Basic paragraph wrapping, escape HTML chars if needed in real use
            html_content += f"    <p>{line}</p>\n"
        html_content += """</body>
</html>
"""
        try:
            with open(full_path, 'w', encoding='utf-8') as f:
                f.write(html_content)
            print("HTML document saved successfully.")
        except IOError as e:
            print(f"Error saving HTML document {full_path}: {e}")

# --- Creator Interface ---
class DocumentProcessor(ABC):
    """Abstract Creator: Declares the factory method."""

    @abstractmethod
    def create_document(self, title: str, content: List[str]) -> Document:
        """The Factory Method."""
        pass

    def process_and_save(self, title: str, content: List[str], output_dir: str, filename_base: str):
        """Processes data and uses the factory method to save the document."""
        # Ensure output directory exists
        try:
            os.makedirs(output_dir, exist_ok=True)
        except OSError as e:
            print(f"Error creating directory {output_dir}: {e}")
            return

        # Create the document using the factory method
        doc = self.create_document(title=title, content=content)

        # Construct the full path without extension
        full_path_base = os.path.join(output_dir, filename_base)

        # Save the document (save method adds the extension)
        print(f"\nProcessing with {self.__class__.__name__}:")
        doc.save(full_path_base)

# --- Concrete Creators ---
class TextProcessor(DocumentProcessor):
    """Concrete Creator: Overrides the factory method for Text documents."""
    def create_document(self, title: str, content: List[str]) -> TextDocument:
        print("TextProcessor: Creating TextDocument.")
        return TextDocument(title, content)

class JSONProcessor(DocumentProcessor):
    """Concrete Creator: Overrides the factory method for JSON documents."""
    def create_document(self, title: str, content: List[str]) -> JSONDocument:
        print("JSONProcessor: Creating JSONDocument.")
        return JSONDocument(title, content)

class HTMLProcessor(DocumentProcessor):
    """Concrete Creator: Overrides the factory method for HTML documents."""
    def create_document(self, title: str, content: List[str]) -> HTMLDocument:
        print("HTMLProcessor: Creating HTMLDocument.")
        return HTMLDocument(title, content)
